RecoEgamma/EgammaTools/plugins/EGRegressionModifierDRN.cc

```diff
@@ -0,0 +1,243 @@
+    gedPhotons_->particles = evt.get(gedPhotons_->token);
+    gedPhotons_->corrections = evt.get(gedPhotons_->correctionsToken);
+  }
+}
+
+void EGRegressionModifierDRN::setEventContent(const edm::EventSetup& iSetup) {}
+
+void EGRegressionModifierDRN::modifyObject(reco::GsfElectron& ele) const {
+  if (!gsfElectrons_)
+    return;
+
+  const std::pair<float, float>& correction = gsfElectrons_->getCorrection(ele);
+
+  if (correction.first <= 0)
+    return;
+
+  ele.setCorrectedEcalEnergy(correction.first, true);
+  ele.setCorrectedEcalEnergyError(correction.second);
+
+  throw cms::Exception("EGRegressionModifierDRN")
+      << "Electron energy corrections not fully implemented yet:" << std::endl
+      << "Still need E/p combination" << std::endl
+      << "Do not enable DRN for electrons" << std::endl;
+
+  const std::pair<float, float> trackerCombo(1.0, 1.0);  //TODO: compute E/p combination
+  const math::XYZTLorentzVector newP4 = ele.p4() * trackerCombo.first / ele.p4().t();
+  ele.correctMomentum(newP4, ele.trackMomentumError(), trackerCombo.second);
+}
+
+void EGRegressionModifierDRN::modifyObject(pat::Electron& ele) const {
+  if (!patElectrons_)
+    return;
+
+  const std::pair<float, float>& correction = patElectrons_->getCorrection(ele);
+
+  if (correction.first <= 0)
+    return;
+
+  ele.setCorrectedEcalEnergy(correction.first, true);
+  ele.setCorrectedEcalEnergyError(correction.second);
+
+  throw cms::Exception("EGRegressionModifierDRN")
+      << "Electron energy corrections not fully implemented yet:" << std::endl
+      << "Still need E/p combination" << std::endl
+      << "Do not enable DRN for electrons" << std::endl;
+
+  const std::pair<float, float> trackerCombo(1.0, 1.0);  //TODO: compute E/p combination
+  const math::XYZTLorentzVector newP4 = ele.p4() * trackerCombo.first / ele.p4().t();
+  ele.correctMomentum(newP4, ele.trackMomentumError(), trackerCombo.second);
+}
+
+void EGRegressionModifierDRN::modifyObject(pat::Photon& pho) const {
+  if (!patPhotons_)
+    return;
+  const std::pair<float, float>& correction = patPhotons_->getCorrection(pho);
+
+  if (correction.first <= 0)
+    return;
+
+  pho.setCorrectedEnergy(pat::Photon::P4type::regression2, correction.first, correction.second, true);
+}
+
+void EGRegressionModifierDRN::modifyObject(reco::Photon& pho) const {
+  if (!gedPhotons_)
+    return;
+
+  const std::pair<float, float>& correction = gedPhotons_->getCorrection(pho);
+
+  if (correction.first <= 0)
+    return;
+
+  pho.setCorrectedEnergy(reco::Photon::P4type::regression2, correction.first, correction.second, true);
+};
+
+template <typename T>
+const std::pair<float, float> EGRegressionModifierDRN::partVars<T>::getCorrection(T& part) const {
+  const math::XYZTLorentzVectorD& partP4 = part.p4();
+
+  bool matched = false;
+  unsigned i;
+  for (i = 0; i < particles.size(); ++i) {
+    const T& partIter = particles.at(i);
+    const auto& p4Iter = partIter.p4();
+    if (p4Iter == partP4) {
```

Comment:
What I'm trying to do with this piece of code is find the ref to the object being modified so that I can index the ValueMap with it and get the correction values. I couldn't figure out how to obtain the ref directly from the object itself, necessitating this loop. I think it should always happen that we can find the exact object being modified in the particles vector, so the check for exact equivalence of their four-vectors should be robust. I have never observed this to fail in all my testing. All that said, I would love to learn of a better way to obtain the ref for an object without needing this loop.

Reply:
Comparing two vectors of doubles is not a good way to check for equivalence. Even if it works on 100 events on your machine, who can say what will happen in the future, on large stats, on other architectures / compilers...
Moreover, isn't it slow to compare several doubles, if in principle just an integer index needs to be compared?

Reply:
I understand that in general comparing doubles is not robust. In this case, the reason I believe it to be robust is that by construction the object `part` came from the collection `particles`. I therefore expect that there will always be an object in `particles` which is exactly equal to `part`, and this loop is just my attempt at finding it (and fwiw I have tested on 10s of thousands of events, not just a hundred). I agree that comparing floats is also much slower than comparing ints. However, I don't start off knowing the integer index of `part` within `particles`, and the intention of this block of code is exactly to find this integer index. What is the better way to do this?

Reply:
Perhaps instead of storing the corrections as a vector, you can use and AssociationMap: https://github.com/cms-sw/cmssw/blob/05bf287fcdb7188310b9dbbd132b0f4df5733a0a/DataFormats/Common/interface/AssociationMap.h

Reply:
Indeed in general it is not possible to construct `edm::Ref`/`edm::Ptr` from just an element of a collection (and an index). In this case it is not even possible to change the function signature to take `Ref`/`Ptr`, because the intention appears to modify the argument object, and `Ref`/`Ptr` give only `const` access.

Assuming this class gets used in conjunction with `pat::ObjectModifier<T>` (like `PATPhotonSlimmer`), are there any guarantees that this loop over the modifiers
https://github.com/cms-sw/cmssw/blob/ce4f5b212642a02e5c7ccff8403fdbf6f0306865/PhysicsTools/PatAlgos/interface/ObjectModifier.h#L26-L29
would not modify the `part.p4()` before calling this object of this class?

In case of `PATPhotonSlimmer` would, in principle, be possible to pass the "input" `edm::Ptr<pat::Photon>` along the `pat::Photon` to be modified that could be used here as an index to the `edm::ValueMap`, but interface-wise that could be annoying for those users of the `ModifyObjectValueFactory` infrastructure that create the objects-to-be-modified on the fly, although those wouldn't be able to use this kind of class anyway.

Reply:
That's a really good point about the potential of the 4-vector being modified in one of the other modifiers. I had not considered that. I'm highly reluctant to change the interface, as this class has been designed to function as a drop-in replacement for the EGRegressionModifierV3, and ideally it should work anywhere that these sorts of modifiers are used. 

For what it's worth, it looks like the way this problem was solved in the [EGExtraInfoModifierFromValueMaps](https://github.com/cms-sw/cmssw/blob/master/RecoEgamma/EgammaTools/interface/EGExtraInfoModifierFromValueMaps.h) was to store the index as a member variable and just increment it in each modifyObject() call. This had struck me as not very robust as it assumes that the calling logic loops through the entire parent collection exactly once and in the correct order, but maybe this is a better assumption than my assumption that the p4() will not be changed by any other modifier.

Reply:
For a possible redesign of the interface (which, if done, I agree would be best done outside of this PR), perhaps a reasonable compromise would be to add [`edm::ElementID`](https://github.com/cms-sw/cmssw/blob/master/DataFormats/Provenance/interface/ElementID.h) in the function signature. The `ElementID` is lightweight (holds only the `ProductID` and the index), and should work automatically e.g. in `edm::ValueMap<T>::operator[]()`. I'd imagine creating it for cases that create `pat::Photon` on the fly to be smaller annoyance than having to create `Ref`/`Ptr`.

Reply:
Okay, I think for the purposes of this PR I will change it to determine the index by assuming that the external calling logic loops through the entire parent collection in order exactly once, as is done in the EGExtraInfoModifierFromValueMaps. And then once we are done with this PR maybe we should have a more detailed discussion with EGM (or whoever the relevant group managing this code is) about updating the interface. Does this sound like a good plan to you?